shared/libpng-util/libpng-util.cpp

```diff
@@ -0,0 +1,122 @@
+***
+*** ENIGMA is free software: you can redistribute it and/or modify it under the
+*** terms of the GNU General Public License as published by the Free Software
+*** Foundation, version 3 of the license or any later version.
+***
+*** This application and its source code is distributed AS-IS, WITHOUT ANY
+*** WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+*** FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+*** details.
+***
+*** You should have received a copy of the GNU General Public License along
+*** with this code. If not, see <http://www.gnu.org/licenses/>
+**/
+
+#include <string>
+
+#include <png.h>
+#include <stdlib.h> // for malloc/free
+
+unsigned libpng_encode32_file(const unsigned char* image, const unsigned w, const unsigned h, const char* filename) {
+  FILE *fp = fopen(filename, "wb");
+
+  png_structp png = png_create_write_struct(PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
+  if (!png) return -1;
+  png_infop info = png_create_info_struct(png);
+  if (!info) {
+    png_destroy_write_struct(&png, NULL);
+    return -2;
+  }
+
+  png_init_io(png, fp);
+
+  // Write header (8 bit colour depth)
+  png_set_IHDR(png, info, w, h,
+               8, PNG_COLOR_TYPE_RGBA, PNG_INTERLACE_NONE,
+               PNG_COMPRESSION_TYPE_BASE, PNG_FILTER_TYPE_BASE);
+
+  png_write_info(png, info);
+
+  for (unsigned i = 0; i < h; ++i) {
+    png_write_row(png, (png_bytep)&image[sizeof(png_byte) * 4 * w * i]);
+  }
+  png_write_end(png, NULL);
+
+  fclose(fp);
+  png_free_data(png, info, PNG_FREE_ALL, -1);
+  png_destroy_write_struct(&png, (png_infopp)NULL);
+
+  return 0;
+}
+
+unsigned libpng_decode32_file(unsigned char** out, unsigned* w, unsigned* h, const char* filename) {
+  (*w) = 0; (*h) = 0;
+  FILE *fp = fopen(filename, "rb");
+
+  png_structp png = png_create_read_struct(PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
+  if (!png) return -1;
+  png_infop info = png_create_info_struct(png);
+  if (!info) {
+    png_destroy_write_struct(&png, NULL);
+    return -2;
+  }
+
+  png_init_io(png, fp);
+  png_read_info(png, info);
+
+  png_byte color_type, bit_depth;
+  (*w)       = png_get_image_width(png, info);
+  (*h)       = png_get_image_height(png, info);
+  color_type = png_get_color_type(png, info);
+  bit_depth  = png_get_bit_depth(png, info);
+
+  // Read any color_type into 8bit depth, RGBA format.
+  // See http://www.libpng.org/pub/png/libpng-manual.txt
+
+  if (bit_depth == 16)
+    png_set_strip_16(png);
+
+  if (color_type == PNG_COLOR_TYPE_PALETTE)
+    png_set_palette_to_rgb(png);
+
+  // PNG_COLOR_TYPE_GRAY_ALPHA is always 8 or 16bit depth.
+  if (color_type == PNG_COLOR_TYPE_GRAY && bit_depth < 8)
+    png_set_expand_gray_1_2_4_to_8(png);
+
+  if (png_get_valid(png, info, PNG_INFO_tRNS))
+    png_set_tRNS_to_alpha(png);
+
+  // These color_type don't have an alpha channel then fill it with 0xff.
+  if (color_type == PNG_COLOR_TYPE_RGB ||
```

Comment:
LibPNG has utilities that can expand color palettes for you, etc. This isn't something the engine needs to worry about (at least not right now). I gave you a utility class to handle all kinds of stuff like this a hundred years ago.

Reply:
What? Where? I don't recall this alleged utility class you provided.

Reply:
https://pastebin.com/9xHptLL4

Reply:
Ok, maybe I do vaguely remember it now. Looking over your code, I have some issues with it. You allocate row pointers like a lot of the other examples, I _hate_ that. If you look at my version, you'll clearly see I only do a single allocation for encoding and decoding. I was doing it your way when my benchmarks were performing _worse_, so I am not inclined to want to go back to doing it that way.

Reply:
Row pointers are how PNG works, Robert; you're just asking it for one at a time instead of all of them at once, for whatever reason, and it's costing you effort, because if you don't care about supporting images that don't fit in a contiguous block of memory (which you're forcing here, anyway), you can just ask PNG to read the whole image and normalize it for you instead of calling `png_set_gray_to_rgb`, etc.

But whatever, I guess. You can address bugs about unsupported PNG formats as they come in. 😛

Reply:
Alright, so I thought a little about it and why I was trying to force single allocation. It was prior to me eliminating malloc/free from your earlier comments. I was concerned that it would cause problems being integrated with `image_formats.cpp` and lead to memory leaks resulting from the inconsistent deletion requirements when using a non-contiguous allocation. Now that I've thought about it, you might actually have a point though we might be able to simplify this even further yet.

However, considering the time I've spent just getting us switched to libpng, and the cost of investing in doing that further right now, I have decided I would like to merge this now anyway. We can save additional improvements and optimizations for down the road. This is completely better than the setup we had with LodePNG and I really don't want to invest more time in this right now while letting this go stale.